credentials/apps/core/models.py

```diff
@@ -192,15 +194,31 @@ def access_token(self):
 
         if not access_token:
             url = f'{self.oauth2_provider_url}/access_token'
-            access_token, expiration_datetime = EdxRestApiClient.get_oauth_access_token(
-                url,
-                self.oauth2_client_id,
-                self.oauth2_client_secret,
-                token_type='jwt'
-            )
-
-            expires = (expiration_datetime - datetime.datetime.utcnow()).seconds
-            cache.set(key, access_token, expires)
+            retries = [15, 30, 45]  # Retry delays in seconds
+            attempt = 0
+
+            while attempt < len(retries) + 1:
+                try:
+                    log.info(f'Feching access token for URL: {url}')
+                    access_token, expiration_datetime = EdxRestApiClient.get_oauth_access_token(
+                        url,
+                        self.oauth2_client_id,
+                        self.oauth2_client_secret,
+                        token_type='jwt'
+                    )
+                    expires = (expiration_datetime - datetime.datetime.utcnow()).seconds
+                    cache.set(key, access_token, expires)
+                    return access_token
+                except Exception as e:
+                    log.info(f'Getting exception wile getting token {e}')
+                    if isinstance(e.response, Response) and e.response.status_code == 403:
+                        attempt += 1
+                        if attempt > len(retries):
```

Comment:
All suggestions are really good and we can address most of them. Thank you for pointing them out. 
As per our discussion with @ziafazal and @taimoor-ahmed-1, unfortunately we can not increase the rate limit since its already 120/m. 

Issue happens when jobs run in parallel and it reaches the rate limit. So we have decided to optimize the jobs; add token cache in jobs and improve the logic so the there should be less requests for token.
This job already has cache also it is hitting token API only once for one site. 

Another suggestion was to create a universal token for worker which can be used for every site, @taimoor-ahmed-1 was looking into the feasibility of that.